### PR TITLE
Introduce Type::getConstantArrays as successor for TypeUtils::getOldContantArrays

### PR DESCRIPTION
**Pull request by herndlm**
and deprecate the now both unused `TypeUtils::getOldContantArrays()` and `TypeUtils::getConstantArrays()`

### CODE REVIEW

File: src/Type/TypeUtils.php [hunk 1]
```diff
@@ -63,6 +63,8 @@ public static function getArrays(Type $type): array
 
 	/**
 	 * @return ConstantArrayType[]
+	 *
+	 * @deprecated Use PHPStan\Type\Type::getConstantArrays() instead and properly deal with optional keys.
```
Comment (herndlm) [inline src/Type/TypeUtils.php]: a bit aggressive maybe? :) this method is not used anymore in phpstan itself btw.

File: src/Type/TypeUtils.php [hunk 2]
```diff
@@ -182,15 +184,6 @@ public static function getEnumCaseObjects(Type $type): array
...
 
-	/**
-	 * @internal
-	 * @return ConstantArrayType[]
-	 */
-	public static function getOldConstantArrays(Type $type): array
-	{
-		return self::map(ConstantArrayType::class, $type, false);
-	}
-
```
Comment (herndlm) [inline src/Type/TypeUtils.php]: the BC checker does not like the removal, which makes sense I guess. should it stay then and additionally be deprecated I assume?

Comment (ondrejmirtes) [reply src/Type/TypeUtils.php]: The BC checker isn't aware of `@internal`. Yeah, probably better to leave the method in the class, I'll remove it in 2.0 :) Also you can add `@deprecated` to it too so I can find it. THanks :)

**Comment by herndlm** [comment]
I was thinking about adding a `NonConstantArrayTypeTrait` and use that in `JustNullableTypeTrait` and e.g. `NonIterableTypeTrait` which covers many types already. But I wasn't sure if that idea is really good and wanted to avoid making it overly complex.

**Comment by herndlm** [comment]
Not sure how and why but it reports https://github.com/phpstan/phpstan-src/blob/0645439ef3ebebc87a85ed8046e73302cbca5bd0/tests/PHPStan/Levels/data/arrayDimFetches.php#L28 now as non-existant offset, additionally to L21 which was already reported and I think it is legit. Updated the levels json file.

**Comment by herndlm** [comment]
No, there are still problems with the levels test. Could be the intersection that was previously not considered via TypeUtils::map I assume. I'll have to check

**Comment by herndlm** [comment]
The failure now is unrelated, bad timing from my side too I guess. Looks like it's related to the composer fork integration test.

**Comment by ondrejmirtes** [comment]
I'm eager to merge this :)

**Comment by herndlm** [comment]
great, 1.9.x might also be a good chance to deprecate those methods :) we just have to be careful with every new usage of deprecated methods that gets added in 1.8.x which could cause phpstan errors on 1.9.x then I guess.

that should not happen often but e.g. https://github.com/phpstan/phpstan-src/pull/1795 for example. I can rebase after that one got merged up. or just fix it afterwards, it's not a big deal I guess..

**Comment by ondrejmirtes** [comment]
Yes, that will definitely happen :)

**Comment by ondrejmirtes** [comment]
Thank you.

**Comment by herndlm** [comment]
I'll deprecate `TypeUtils::getArrays()` and `TypeUtils::getAnyArrays()` via https://github.com/phpstan/phpstan-src/pull/1687 a bit later today

**Pull request merged**